ONSdigital/eq-questionnaire-runner
# Supplementary data dependencies fix

### What is the context of this PR?
This PR replaces [an old PR](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1224) that implemented a fix to progress not updating on supplementary data change using the new split dependencies. The other branch will be deleted upon the merge of this one.

It fixes the following issues:

1) If a question depends on the length of a supplementary list, and a new supplementary dataset is used with a list item count that opens up the question, previously the section containing the question would remain completed. It now reverts to in progress. This is true when the supplementary dataset has added or removed items, both will work.

2) If a section contains a list collector with repeating blocks depending on supplementary data. If the supplementary data changes and a list item is removed, the section will revert to in progress. This includes for repeating blocks which will require revisiting.

### How to review
The questionnaire store updater has been split into a base updater which does not have location and question, and the main questionnaire store updater now inherits from this. 

**To Make review easier, for now, the two questionnaire store updaters reside in the same file, this way the reviewer can see which methods have been removed from the base updater, and added to the new class**

After merge, a follow up will move it to its own file

### How to test functionally
Run the mock SDS locally on [the branch with an additional dataset](https://github.com/ONSdigital/eq-runner-mock-sds/pull/3)
Use the `test_supplementary_data` schema, and try swapping between the three datasets while clicking the use previous response ID button.

- If using the first supplementary dataset with 2 items, if you complete the production targets section and relaunch the survey with the second supplementary dataset with 3 items, the section should become incomplete.

- If using the second supplementary dataset with 3 items, if you complete the sales targets section, then relaunch the survey using the first supplementary dataset which has 2 items, you should see the section revert to incomplete

- If the product details section is completed whilst using the second supplementary dataset, when moving to the first supplementary dataset which contains fewer items, the section should revert to in progress and force you to reconfirm the calculated summaries within and fill incomplete repeating blocks.

- If you have completed the repeating section for the employees on dataset 1 or dataset 2 and you then use the 3rd dataset, you should see every previously completed repeat revert to in progress, because each repeat now has a new question which was unlocked by having 3 supplementary employees. (Any employees that exist only in the 3rd dataset will appear 'not started' since that section is new)

### Checklist
